iroh-bytes/src/store/traits.rs

```diff
@@ -428,6 +433,141 @@ pub trait Store: ReadableStore + MapMut {
+
+    /// Validate the database
+    fn validate(
+        &self,
+        repair: bool,
+        tx: BoxedProgressSender<ValidateProgress>,
+    ) -> impl Future<Output = io::Result<()>> + Send {
+        validate_impl(self, repair, tx)
+    }
+}
+
+async fn validate_impl(
+    store: &impl Store,
+    repair: bool,
+    tx: BoxedProgressSender<ValidateProgress>,
+) -> io::Result<()> {
+    let validate_parallelism: usize = num_cpus::get();
+    let lp = LocalPoolHandle::new(validate_parallelism);
+    let complete = store.blobs().await?.collect::<io::Result<Vec<_>>>()?;
+    let partial = store
+        .partial_blobs()
+        .await?
+        .collect::<io::Result<Vec<_>>>()?;
+    tx.send(ValidateProgress::Starting {
+        total: complete.len() as u64,
+    })
+    .await?;
+    let complete_result = futures::stream::iter(complete)
+        .map(|hash| {
+            let store = store.clone();
+            let tx = tx.clone();
+            lp.spawn_pinned(move || async move {
+                let entry = store
+                    .get(&hash)
+                    .await?
+                    .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "entry not found"))?;
+                let size = entry.size().value();
+                let outboard = entry.outboard().await?;
+                let data = entry.data_reader().await?;
+                let chunk_ranges = ChunkRanges::all();
+                let mut ranges = bao_tree::io::fsm::valid_ranges(outboard, data, &chunk_ranges);
+                let id = tx.new_id();
+                tx.send(ValidateProgress::Entry {
+                    id,
+                    hash,
+                    path: None,
+                    size,
+                })
+                .await?;
+                let mut actual_chunk_ranges = ChunkRanges::empty();
+                while let Some(item) = ranges.next().await {
+                    let item = item?;
+                    let offset = item.start.to_bytes().0;
+                    actual_chunk_ranges |= ChunkRanges::from(item);
+                    tx.try_send(ValidateProgress::EntryProgress { id, offset })?;
+                }
+                let expected_chunk_range =
+                    ChunkRanges::from(..BaoTree::new(ByteNum(size), IROH_BLOCK_SIZE).chunks());
+                let incomplete = actual_chunk_ranges == expected_chunk_range;
+                let error = if incomplete {
+                    None
+                } else {
+                    Some(format!(
+                        "expected chunk ranges {:?}, got chunk ranges {:?}",
+                        expected_chunk_range, actual_chunk_ranges
+                    ))
+                };
+                tx.send(ValidateProgress::EntryDone { id, error }).await?;
+                drop(ranges);
+                drop(entry);
+                io::Result::Ok((hash, incomplete))
+            })
+        })
+        .buffer_unordered(validate_parallelism)
```

Comment:
sorry, but no, this will segfault iroh again..

Reply:
Do you have an issue for that?

Reply:
https://github.com/n0-computer/iroh/issues/1646 all things are linked there